### PR TITLE
implement check_type and fix different errors

### PR DESCRIPTION
**Pull request by markuman**
##### SUMMARY
closes: #19

* Implement `check_type` feature.
* fix different other issues
  * `reader_hostgroup: 0`
  * ability to change `reader_hostgroup` in general  
  * add default value for `comment`. `""` is used in source code

##### ISSUE TYPE

- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
proxysql_replication_hostgroups

### CODE REVIEW

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 1]
```diff
@@ -115,18 +126,15 @@ def perform_checks(module):
...
 
-    if not module.params["reader_hostgroup"] == \
-            module.params["writer_hostgroup"]:
-        if not module.params["reader_hostgroup"] > 0:
-            module.fail_json(
-                msg=("writer_hostgroup must be a integer greater than" +
-                     " or equal to 0")
-            )
-    else:
+    if not module.params["reader_hostgroup"] >= 0:
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:129]:
```suggestion
    if module.params["reader_hostgroup"] < 0:
```

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 2]
```diff
@@ -115,18 +126,15 @@ def perform_checks(module):
...
         module.fail_json(
-            msg="reader_hostgroup cannot equal writer_hostgroup"
+            msg="reader_hostgroup must be a integer greater than or equal to 0"
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:131]:
```suggestion
            msg="reader_hostgroup must be an integer greater than or equal to 0"
```

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 3]
```diff
@@ -278,17 +283,60 @@ def delete_repl_group(self, check_mode, result, cursor):
                              " mysql_replication_hostgroups, however" +
                              " check_mode is enabled.")
 
+    def update_check_type(self, cursor):
+        try:
+            query_string = \
+                """UPDATE mysql_replication_hostgroups
+                    SET check_type = %s
+                    WHERE writer_hostgroup = %s
+                """
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:292]:
```suggestion
            query_string = ("UPDATE mysql_replication_hostgroups "
                            "SET check_type = %s "
                            "WHERE writer_hostgroup = %s")
```
I suggest using this format. It looks more compact and efficient in terms of passing less whitespaces.

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 4]
```diff
@@ -248,23 +238,38 @@ def create_repl_group(self, check_mode, result, cursor):
...
-        if not check_mode:
+        if int(current.get('reader_hostgroup')) != self.reader_hostgroup:
+            result['changed'] = True
+            result['msg'] = "Updated replication hostgroups in check_mode"
+            if not self.check_mode:
+                result['msg'] = "Updated replication hostgroups"
+                self.update_reader_hostgroup(cursor)
+
+        result['repl_group'] = \
+            self.get_repl_group_config(cursor)
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:266]:
```suggestion
        result['repl_group'] = self.get_repl_group_config(cursor)
```

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 5]
```diff
@@ -278,17 +283,60 @@ def delete_repl_group(self, check_mode, result, cursor):
...
+        query_string = \
+            """UPDATE mysql_replication_hostgroups
+                SET comment = %s
+                WHERE writer_hostgroup = %s
+            """
+
+        query_data = \
+            [self.comment, self.writer_hostgroup]
+
+        cursor.execute(query_string, query_data)
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:323]:
```suggestion
        cursor.execute(query_string, (self.comment, self.writer_hostgroup))
```
just a suggestion. imo the function invocation is still pretty short. though it's a question of personal taste

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 6]
```diff
@@ -278,17 +283,60 @@ def delete_repl_group(self, check_mode, result, cursor):
...
     argument_spec = proxysql_common_argument_spec()
     argument_spec.update(
         writer_hostgroup=dict(required=True, type='int'),
         reader_hostgroup=dict(required=True, type='int'),
-        comment=dict(type='str'),
+        check_type=dict(type='str', default='read_only', choices=["read_only",
+                                                                  "innodb_read_only",
+                                                                  "super_read_only",
+                                                                  "read_only|innodb_read_only",
+                                                                  "read_only&innodb_read_only"]),
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:338]:
```suggestion
        check_type=dict(type='str', default='read_only', choices=['read_only',
                                                                  'innodb_read_only',
                                                                  'super_read_only',
                                                                  'read_only|innodb_read_only',
                                                                  'read_only&innodb_read_only']),
```
i would suggest using a common quote style in adjacent pieces of code

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 7]
```diff
@@ -318,31 +366,24 @@ def main():
...
-                    result['msg'] = ("The repl group already exists in" +
-                                     " mysql_replication_hostgroups and" +
-                                     " doesn't need to be updated.")
-                    result['repl_group'] = \
-                        proxysql_repl_group.get_repl_group_config(cursor)
+                proxysql_repl_group.update_repl_group(result,
+                                                      cursor)
+
+                result['repl_group'] = \
+                    proxysql_repl_group.get_repl_group_config(cursor)
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:386]:
```suggestion
                proxysql_repl_group.update_repl_group(result, cursor)

                result['repl_group'] = proxysql_repl_group.get_repl_group_config(cursor)
```

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 8]
```diff
@@ -353,8 +394,7 @@ def main():
         try:
             if proxysql_repl_group.check_repl_group_config(cursor,
                                                            keys=True):
-                proxysql_repl_group.delete_repl_group(module.check_mode,
-                                                      result,
+                proxysql_repl_group.delete_repl_group(result,
                                                       cursor)
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:398]:
```suggestion
                proxysql_repl_group.delete_repl_group(result, cursor)
```

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 9]
```diff
@@ -278,17 +283,60 @@ def delete_repl_group(self, check_mode, result, cursor):
...
+            cursor.execute(query_string, query_data)
+        except Exception as e:
+            pass
+
+    def update_reader_hostgroup(self, cursor):
+        query_string = \
+            """UPDATE mysql_replication_hostgroups
+                SET reader_hostgroup = %s
+                WHERE writer_hostgroup = %s
+            """
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:306]:
```suggestion
        query_string = ("UPDATE mysql_replication_hostgroups "
                        "SET reader_hostgroup = %s "
                        "WHERE writer_hostgroup = %s")
```

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 10]
```diff
@@ -278,17 +283,60 @@ def delete_repl_group(self, check_mode, result, cursor):
...
+
+        query_data = \
+            [self.reader_hostgroup, self.writer_hostgroup]
+
+        cursor.execute(query_string, query_data)
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:311]:
```suggestion
        cursor.execute(query_string, (self.reader_hostgroup, self.writer_hostgroup))
```

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 11]
```diff
@@ -278,17 +283,60 @@ def delete_repl_group(self, check_mode, result, cursor):
...
+            query_string = \
+                """UPDATE mysql_replication_hostgroups
+                    SET check_type = %s
+                    WHERE writer_hostgroup = %s
+                """
+
+            query_data = \
+                [self.check_type, self.writer_hostgroup]
+
+            cursor.execute(query_string, query_data)
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:297]:
```suggestion
            cursor.execute(query_string, (self.check_type, self.writer_hostgroup))
```

**Review by Andersson007 — COMMENTED** [review]
@markuman thanks for the new feature and the fixes!

My suggestions are
1. If the PR contains bugfixes, it would be good to add `bugfixes:` in the changelog fragment
2. in the future, it's better to add not tightly related things with separate PRs because a) it's much easier to review b) it's easier to revert changes later if needed
3. several cosmetic not necessary suggestions
4. idea: would be perfect to start checking actual system state with `shell:` or `command:` modules in integration tests in the future, i.e. to query the test DB via cli tools OR `proxysql_query:` module that we discussed should definitely help (maybe `proxysql_info` covers this case?).
5. it would be nice to have an example for the new option in the EXAMPLES block

Thank you!

**Comment by markuman** [comment]
@Andersson007 

> in the future, it's better to add not tightly related things with separate PRs because a) it's much easier to review b) it's easier to revert changes later if needed

Absolutely right. It just happen in a flow of coding.  
I can split it if you like.

> idea: would be perfect to start checking actual system state with shell: or command: modules in integration tests in the future, i.e. to query the test DB via cli tools OR proxysql_query: module that we discussed should definitely help (maybe proxysql_info covers this case?).

You mean, we did it now with shell/command and replace it with `proxysql_query` later?  
I think it is also possible with `proxysql_info` now.

**Comment by Andersson007** [comment]
@markuman thanks for working on this!
- I think no need to split it now (as you prefer though)

> You mean, we did it now with shell/command and replace it with proxysql_query later?
I think it is also possible with proxysql_info now.

Ah, yes, i greped through the tests/integration directory and i see them used.
Though I don't see them in the integration tests for this PR. Not sure if it's applicable. I personally prefer to check every change with another module. In postgres / mysql we have the`*_query` module which helps much.

**Comment by Andersson007** [comment]
if the check is not applicable here, feel free to skip my suggestion

**Review by Andersson007 — APPROVED** [review]
LGTM

**Comment by markuman** [comment]
Just one small typo change https://github.com/ansible-collections/community.proxysql/pull/69/commits/bb59a10a18361c75b3ee6d374024da8a9f18ca55

**Comment by markuman** [comment]
And one left query style: https://github.com/ansible-collections/community.proxysql/pull/69/commits/f17af881d09e2f942a71972d7969d84d053a84fe

**Review by Andersson007 — APPROVED** [review]
Thanks!

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 12]
```diff
@@ -248,23 +248,38 @@ def create_repl_group(self, check_mode, result, cursor):
...
+    def update_repl_group(self, result, cursor):
+        current = self.get_repl_group_config(cursor)
+
+        if self.check_type_support:
+            if current.get('check_type') != self.check_type:
+                result['changed'] = True
+                result['msg'] = "Updated replication hostgroups in check_mode"
+                if not self.check_mode:
+                    result['msg'] = "Updated replication hostgroups"
+                    self.update_check_type(cursor)
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:260]:
```suggestion
                if not self.check_mode:
                    result['msg'] = "Updated replication hostgroups"
                    self.update_check_type(cursor)
                else:
                    result['msg'] = "Updated replication hostgroups in check_mode"
```
I can see that the result will be the same anyway. Just a suggestion:) A question of personal preferences

File: plugins/modules/proxysql_replication_hostgroups.py [hunk 13]
```diff
@@ -140,28 +158,25 @@ def load_config_to_runtime(cursor):
...
-    def __init__(self, module):
+    def __init__(self, module, version):
         self.state = module.params["state"]
         self.save_to_disk = module.params["save_to_disk"]
         self.load_to_runtime = module.params["load_to_runtime"]
         self.writer_hostgroup = module.params["writer_hostgroup"]
         self.reader_hostgroup = module.params["reader_hostgroup"]
         self.comment = module.params["comment"]
+        self.check_type = module.params["check_type"]
+        self.check_type_support = version.get('major') >= 2
```
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:169]:
```suggestion
        self.check_type_support = version.get('major') >= 2
```
cat there be cases when users use 2.0.0? We are stating that the minimal version supporting this feature is 2.0.1

Comment (markuman) [reply plugins/modules/proxysql_replication_hostgroups.py:169]: Good question.  
The previous check made `check_type` functionality fail with proxysql 2.2.0.  
furthermore, there is not 2.0.0 release: https://github.com/sysown/proxysql/releases/tag/v2.0.0  
The first proxysql 2.x.x release is 2.0.1

Comment (markuman) [reply plugins/modules/proxysql_replication_hostgroups.py:169]: So it is sufficient to check the major version

File: plugins/modules/proxysql_replication_hostgroups.py (hunk 12 above)
Comment (Andersson007) [inline plugins/modules/proxysql_replication_hostgroups.py:260]:
```suggestion
        if self.check_type_support and current.get('check_type') != self.check_type:
            result['changed'] = True
            if not self.check_mode:
                result['msg'] = "Updated replication hostgroups"
                self.update_check_type(cursor)
            else:
                result['msg'] = "Updated replication hostgroups in check_mode"
```
to decrease nesting a bit

File: tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml [hunk 14]
```diff
@@ -82,6 +82,120 @@
...
+        login_user: admin
+        login_password: admin
+        writer_hostgroup: 100
+        reader_hostgroup: 300
+        check_type: innodb_read_only
+      register: special_condition
+      check_mode: yes
+
+    - name: verify check_mode change for proxysql > 2
+      when: proxysql_information.version.major == 2
```
Comment (Andersson007) [inline tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml:131]:
```suggestion
      when: proxysql_information.version.major >= 2
```
makes sense?

File: tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml [hunk 15]
```diff
@@ -82,6 +82,120 @@
...
+      community.proxysql.proxysql_replication_hostgroups:
+        login_user: admin
+        login_password: admin
+        writer_hostgroup: 100
+        reader_hostgroup: 300
+        check_type: innodb_read_only
+      register: special_condition
+
+    - name: verify change for proxysql > 2
+      when: proxysql_information.version.major == 2
```
Comment (Andersson007) [inline tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml:146]:
```suggestion
      when: proxysql_information.version.major >= 2
```

how about adding a check `when: proxysql_information.version.major < 2` ? Can we check, e.g., that `- special_condition is not changed`?

File: tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml [hunk 16]
```diff
@@ -82,6 +82,120 @@
...
+      community.proxysql.proxysql_replication_hostgroups:
+        login_user: admin
+        login_password: admin
+        writer_hostgroup: 100
+        reader_hostgroup: 300
+        check_type: innodb_read_only
+      register: special_condition
+
+    - name: verify idempotent change for proxysql > 2
+      when: proxysql_information.version.major == 2
```
Comment (Andersson007) [inline tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml:161]:
```suggestion
    - name: verify idempotent change for proxysql >= 2
      when: proxysql_information.version.major >= 2
```

File: tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml (hunk 15 above)
Comment (markuman) [reply tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml:146]: Yes. I will add

**Comment by Andersson007** [comment]
I'm going to checkout and run the tests locally in 5-10 minutes.

**Comment by markuman** [comment]
> I'm going to checkout and run the tests locally in 5-10 minutes.

I will run now against various version, included 1.4.15

File: tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml (hunk 15 above)
Comment (Andersson007) [reply tests/integration/targets/test_proxysql_replication_hostgroups/tasks/main.yml:152]: how about changing the line as suggested same as you did a bit above?:)

**Comment by Andersson007** [comment]
everything passed locally

**Review by Andersson007 — APPROVED** [review]
LGTM

**Pull request merged**